app-ios/tutanota/Sources/Contacts/IosMobileContactsFacade.swift

```diff
@@ -0,0 +1,577 @@
+private typealias TutaToStructuredContacts = [String: StructuredContact]
+
+private let CONTACTS_MAPPINGS = "ContactsMappings"
+
+struct UserContactMapping: Codable {
+	let username: String
+	var systemGroupIdentifier: String
+	var localContactIdentifierToServerId: [String: String]
+	var localContactIdentifierToHash: [String: Int]
+}
+
+private let ALL_SUPPORTED_CONTACT_KEYS: [CNKeyDescriptor] =
+	[
+		CNContactIdentifierKey, CNContactGivenNameKey, CNContactFamilyNameKey, CNContactNicknameKey, CNContactOrganizationNameKey, CNContactBirthdayKey,
+		CNContactEmailAddressesKey, CNContactPhoneNumbersKey, CNContactPostalAddressesKey,
+	] as [CNKeyDescriptor]
+
+/// Handles synchronization between contacts in Tuta and contacts on the device.
+class IosMobileContactsFacade: MobileContactsFacade {
+	private let userDefaults: UserDefaults
+
+	init(userDefault: UserDefaults) { self.userDefaults = userDefault }
+
+	func findSuggestions(_ query: String) async throws -> [ContactSuggestion] {
+		try await acquireContactsPermission()
+		return try self.queryContactSuggestions(query: query, upTo: 10)
+	}
+
+	func saveContacts(_ username: String, _ contacts: [StructuredContact]) async throws {
+		try await acquireContactsPermission()
+		var mapping = try self.getOrCreateMapping(username: username)
+		let queryResult = try self.matchStoredContacts(against: contacts, forUser: &mapping)
+		try self.insert(contacts: queryResult.newServerContacts, forUser: &mapping)
+		try self.update(contacts: queryResult.existingServerContacts, forUser: &mapping)
+		for unmappedDeviceContact in queryResult.nativeContactWithoutSourceId {
+			mapping.localContactIdentifierToServerId[unmappedDeviceContact.contact.identifier] = unmappedDeviceContact.serverId
+			mapping.localContactIdentifierToHash[unmappedDeviceContact.localIdentifier] = unmappedDeviceContact.contact
+				.toStructuredContact(serverId: unmappedDeviceContact.serverId).computeHash()
+		}
+
+		// The hash will not match and that's expected as we already returned it as edited on device in sync,
+		// but we need to update those contacts, too, if they are passed to us.
+		try self.update(contacts: queryResult.editedOnDevice, forUser: &mapping)
+
+		self.saveMapping(mapping, forUsername: mapping.username)
+	}
+
+	func syncContacts(_ username: String, _ contacts: [StructuredContact]) async throws -> ContactSyncResult {
+		try await acquireContactsPermission()
+		var mapping = try self.getOrCreateMapping(username: username)
+		let matchResult = try self.matchStoredContacts(against: contacts, forUser: &mapping)
+		try self.insert(contacts: matchResult.newServerContacts, forUser: &mapping)
+		if !matchResult.deletedOnServer.isEmpty { try self.delete(contactsWithServerIDs: matchResult.deletedOnServer, forUser: &mapping) }
+		try self.update(contacts: matchResult.existingServerContacts, forUser: &mapping)
+
+		// For sync it normally wouldn't happen that we have a contact without source/server id but for existing contacts without
```

Comment:
Actually, only during sync we have contacts without origin/server id since the method saveContacts is usually called from the web part with all the information belonging to a contact